### PR TITLE
refactor(ast/expr/planner): unify Span and Result

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Move `Span` to `common_exception`
- Use `common_exception::Result` in `common_expression`
- Remove lifetimes on AST

### CODE REVIEW

**Comment by Xuanwo** [comment]
> * Move `Span` to `common_exception`
> * Use `common_exception::Result` in `common_expression`

Oh, I expect `ast` to have it's own error/result types...

**Comment by andylokandy** [comment]
`ast` indeed has its own Error and Result for parsing, meanwhile, it provides `common_exception::Result` on the outmost API.

**Review by Xuanwo — APPROVED** [review]

**Comment by BohuTANG** [comment]
It's strange:

In others PR:`Dev Linux / test_sqllogic_ydb_cluster_linux (pull_request) Successful in 7m`, but with this PR, it's always timeout.

**Review by PsiACE — APPROVED** [review]

**Comment by BohuTANG** [comment]
> It's strange:
> 
> In other PR(https://github.com/datafuselabs/databend/actions/runs/3997455898/jobs/6859024615):
> 
> ```
> Dev Linux / test_sqllogic_ydb_cluster_linux (pull_request) Successful in 7m
> ```
> 
> , but with this PR, it's always timeout:
> 
> ```
> Dev Linux / test_sqllogic_ydb_cluster_linux (pull_request) Cancelled after 30m
> ```

It's hard to address whether it's the `databend-sqllogictests` params issue or the patch from the github CI output log.
It would be better to show the full command of `databend-sqllogictests` in the CI log.

cc @everpcpc @xudong963

**Comment by everpcpc** [comment]
```
2023-01-25T02:36:55.2951579Z Starting databend-sqllogic tests
2023-01-25T02:36:55.2989816Z MySQL client starts to run...
2023-01-25T02:46:26.5483082Z Run all tests using 571248 ms
2023-01-25T02:46:26.5483303Z Http client starts to run...
2023-01-25T03:02:45.3613614Z Run all tests using 978812 ms
2023-01-25T03:02:45.3613945Z Clickhouse http client starts to run...
2023-01-25T03:06:10.3196061Z ##[error]The operation was canceled.
```
There seems to be some performance issue?

**Comment by andylokandy** [comment]
yes, there is significant performance regression in cluster mode (not found in standalone mode). Still investigating.

**Comment by BohuTANG** [comment]
Try to use https://github.com/datafuselabs/databend/pull/9757 to address the most time cost of `Dev Linux / test_sqllogic_ydb_cluster_linux (pull_request) Cancelled after 30m`

**Comment by BohuTANG** [comment]
@mergify update

File: src/query/expression/src/type_check.rs [hunk 1]
```diff
@@ -283,7 +283,7 @@ impl Subsitution {
         for (idx, ty2) in other.0 {
             if let Some(ty1) = self.0.remove(&idx) {
                 let common_ty = common_super_type(ty2.clone(), ty1.clone()).ok_or_else(|| {
-                    ErrorCode::SemanticError(format!(
```
Comment (sundy-li) [inline src/query/expression/src/type_check.rs]: Backtrace is controlled by env.

```
pub fn capture() -> Backtrace {
      if !Backtrace::enabled() {
          return Backtrace { inner: Inner::Disabled };
      }
      Backtrace::create(Backtrace::capture as usize)
  }
```

Comment (sundy-li) [reply src/query/expression/src/type_check.rs]: But we default to open this in ci tests.

```
# Enable backtrace to debug https://github.com/datafuselabs/databend/issues/7986
export RUST_BACKTRACE=full
./scripts/ci/deploy/databend-query-cluster-3-nodes.sh

SCRIPT_PATH="$(cd "$(dirname "$0")" >/dev/null 2>&1 && pwd)"
cd "$SCRIPT_PATH/../../tests" || exit

echo "Starting databend-test"
# 13_0004_q4: https://github.com/datafuselabs/databend/issues/8107
./databend-test --mode 'cluster' --run-dir 0_stateless --skip '13_0004_q4'

```

Comment (sundy-li) [reply src/query/expression/src/type_check.rs]: The ErrorCode should be lazy initialized unless there are some errors captured in expression framework.

**Pull request merged**